File: src/paths.ts
import { homedir } from "node:os";
import { basename, dirname, join, relative } from "node:path";

export function pathBasename(path: string): string {
  return basename(path);
}

export function pathDirname(path: string): string {
  return dirname(path);
}

export function pathJoin(...parts: string[]): string {
  return join(...parts);
}

export function pathRelative(from: string, to: string): string {
  return relative(from, to);
}

export function getHomeDir(): string {
  return homedir();
}

export function expandHomePath(path: string): string {
  if (!path.startsWith("~")) {
    return path;
  }
  return pathJoin(getHomeDir(), path.replace(/^~\/?/, ""));
}

export function formatRelativeDir(
  targetDir: string,
  cwd: string,
  homeDir: string,
  showRelativeToHome: boolean,
): string {
  if (showRelativeToHome && homeDir && targetDir.startsWith(homeDir)) {
    const relDir = pathRelative(homeDir, targetDir);
    return relDir === "" ? "~/" : `~/${relDir}`;
  }
  const relDir = pathRelative(cwd, targetDir);
  if (!relDir) {
    return "./";
  }
  if (!relDir.startsWith(".")) {
    return `./${relDir}`;
  }
  return `${relDir}/`;
}

File: paths.go
package main

import (
	"os"
	"path/filepath"
	"strings"
)

// pathBasename returns the last element of path
func pathBasename(path string) string {
	return filepath.Base(path)
}

// pathDirname returns all but the last element of path
func pathDirname(path string) string {
	return filepath.Dir(path)
}

// pathJoin joins path elements
func pathJoin(parts ...string) string {
	return filepath.Join(parts...)
}

// pathRelative returns the relative path from 'from' to 'to'
func pathRelative(from, to string) string {
	rel, err := filepath.Rel(from, to)
	if err != nil {
		return to
	}
	return rel
}

// getHomeDir returns the user's home directory
func getHomeDir() string {
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	return home
}

// expandHomePath expands ~ in paths to the home directory
func expandHomePath(path string) string {
	if !strings.HasPrefix(path, "~") {
		return path
	}
	home := getHomeDir()
	if home == "" {
		return path
	}
	if path == "~" {
		return home
	}
	if strings.HasPrefix(path, "~/") {
		return pathJoin(home, path[2:])
	}
	return path
}

// formatRelativeDir formats a directory path relative to cwd or home
func formatRelativeDir(targetDir, cwd, homeDir string, showRelativeToHome bool) string {
	if showRelativeToHome && homeDir != "" && strings.HasPrefix(targetDir, homeDir) {
		relDir := pathRelative(homeDir, targetDir)
		if relDir == "" || relDir == "." {
			return "~/"
		}
		return "~/" + relDir
	}
	relDir := pathRelative(cwd, targetDir)
	if relDir == "" || relDir == "." {
		return "./"
	}
	if !strings.HasPrefix(relDir, ".") {
		return "./" + relDir
	}
	return relDir + "/"
}
